HLTrigger/special/plugins/HLTPixelThrustFilter.cc

```diff
@@ -0,0 +1,107 @@
+  bool hltFilter(edm::Event&,
+                 const edm::EventSetup&,
+                 trigger::TriggerFilterObjectWithRefs& filterproduct) const override;
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> const trackerGeometryRcdToken_;
+  edm::ESGetToken<PixelClusterParameterEstimator, TkPixelCPERecord> const pixelCPERcdToken_;
+  edm::InputTag inputTag_;  // input tag identifying product containing pixel clusters
+  double min_thrust_;       // minimum thrust
+  double max_thrust_;       // maximum thrust
+  edm::EDGetTokenT<edmNew::DetSetVector<SiPixelCluster> > inputToken_;
+};
+
+//
+// constructors and destructor
+//
+
+HLTPixelThrustFilter::HLTPixelThrustFilter(const edm::ParameterSet& config)
+    : HLTFilter(config),
+      trackerGeometryRcdToken_(esConsumes()),
+      pixelCPERcdToken_(esConsumes<PixelClusterParameterEstimator, TkPixelCPERecord>(
+          config.getParameter<edm::ESInputTag>("pixelCPE"))),
+      inputTag_(config.getParameter<edm::InputTag>("inputTag")),
+      min_thrust_(config.getParameter<double>("minThrust")),
+      max_thrust_(config.getParameter<double>("maxThrust")) {
+  inputToken_ = consumes<edmNew::DetSetVector<SiPixelCluster> >(inputTag_);
+}
+
+HLTPixelThrustFilter::~HLTPixelThrustFilter() = default;
+
+void HLTPixelThrustFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+  desc.add<edm::InputTag>("inputTag", edm::InputTag("hltSiPixelClusters"));
+  desc.add<edm::ESInputTag>("pixelCPE", edm::ESInputTag{"", "hltESPPixelCPEGeneric"});
+  desc.add<double>("minThrust", 0);
+  desc.add<double>("maxThrust", 0);
+  descriptions.add("hltPixelThrustFilter", desc);
+}
+
+//
+// member functions
+//
+// ------------ method called to produce the data  ------------
+bool HLTPixelThrustFilter::hltFilter(edm::Event& event,
+                                     const edm::EventSetup& iSetup,
+                                     trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  // All HLT filters must create and fill an HLT filter object,
+  // recording any reconstructed physics objects satisfying (or not)
+  // this HLT filter, and place it in the Event.
+
+  // The filter object
+  if (saveTags())
+    filterproduct.addCollectionTag(inputTag_);
+
+  // get hold of products from Event
+  auto const& clusters = event.get(inputToken_);
+  auto const& trackerGeo = iSetup.getData(trackerGeometryRcdToken_);
+  auto const& pixelCPE = iSetup.getData(pixelCPERcdToken_);
+
+  std::vector<reco::LeafCandidate> vec;
+  for (auto DSViter = clusters.begin(); DSViter != clusters.end(); DSViter++) {
+    auto const& detid = DSViter->detId();
+    auto const& surface = trackerGeo.idToDet(detid)->surface();
+    auto const& detunit = trackerGeo.idToDetUnit(detid);
+    for (auto const& cluster : *DSViter) {
+      auto const& pos = surface.toGlobal(pixelCPE.localParametersV(cluster, *detunit)[0].first);
```

Comment:
for this application do you really need a(n expensive) call to the CPE? 
Would an analog linear average position (barycenter) via `SiPixelCluster::x()` and `SiPixelCluster::y()` be enough?

Reply:
After checking that the precision of the pixel barycenter was good enough, removed the CPE and changed to the barycenter approach in 31fb30855064b1366f53d5483f7d7402e7c6ddc1